zerver/migrations/0382_revoke_invitations_from_deactivated_users.py

```diff
@@ -0,0 +1,77 @@
+from typing import List
+
+from django.db import migrations
+from django.db.backends.postgresql.schema import DatabaseSchemaEditor
+from django.db.migrations.state import StateApps
+from django.utils.timezone import now as timezone_now
+
+
+def revoke_invitations(apps: StateApps, schema_editor: DatabaseSchemaEditor) -> None:
+    Realm = apps.get_model("zerver", "Realm")
+    Confirmation = apps.get_model("confirmation", "Confirmation")
+    Confirmation.INVITATION = 2
+    Confirmation.MULTIUSE_INVITE = 6
+    PreregistrationUser = apps.get_model("zerver", "PreregistrationUser")
+    UserProfile = apps.get_model("zerver", "UserProfile")
+    MultiuseInvite = apps.get_model("zerver", "MultiuseInvite")
+
+    STATUS_ACTIVE = 1
+    STATUS_REVOKED = 2
+
+    def get_valid_invite_confirmations_generated_by_users(
+        user_ids: List[int],
+    ) -> List[int]:
+        prereg_user_ids = (
+            PreregistrationUser.objects.filter(referred_by_id__in=user_ids)
+            .exclude(status__in=[STATUS_ACTIVE, STATUS_REVOKED])
```

Comment:
Why should `STATUS_ACTIVE` be excluded?  I'm not sure that precisely aligns with "No longer active because someone used it" -- I seem to recall things get moved to STATUS_ACTIVE when the user visits the link, regardless of whether they complete account creation, but maybe my memory is off.

Reply:
Yeah, you're right, actually accept STATUS_ACTIVE for creating a new account. I was going by how `filter_to_valid_prereg_users` does, but that function has a bit of a different role. So, we want to revoke STATUS_ACTIVEs in tihs migration as well - fixed that.

But, this is still kind of a mess, because we this weird thing - in `process_new_human_user` we call `revoke_preregistration_users(user_profile, prereg_user, realm_creation)`, which does:
```
def revoke_preregistration_users(
    created_user_profile: UserProfile,
    used_preregistration_user: Optional[PreregistrationUser],
    realm_creation: bool,
) -> None:
    if used_preregistration_user is None:
        assert not realm_creation, "realm_creation should only happen with a PreregistrationUser"

    if used_preregistration_user is not None:
        used_preregistration_user.status = confirmation_settings.STATUS_ACTIVE
        used_preregistration_user.save(update_fields=["status"])
      .......
```

so the used prereguser gets marked here as ACTIVE. This feels like we may want some other status number here, or some other way altogether, to mark the PreregUser was used to create a UserProfile (distinguishing it from the "the confirmation link was clicked" status)?

Reply:
Yes, I agree we should fix that -- probably as part of the PR that also records which user was created by one of these which we want to do anyway.